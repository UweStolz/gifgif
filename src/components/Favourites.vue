<template>
  <div>
    <v-card-title class="display-2 justify-center">
      Favourites
    </v-card-title>

    <v-expansion-panels
      popout
      hover
    >
      <v-expansion-panel
        v-for="x in 5"
        :key="x"
      >
        <v-expansion-panel-header>
          <v-row justify="center">
            <v-icon
              v-for="y in x++"
              :key="y"
              x-large
              color="red"
            >
              {{ icons.mdiHeart }}
            </v-icon>
          </v-row>
        </v-expansion-panel-header>
        <v-expansion-panel-content>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
        </v-expansion-panel-content>
      </v-expansion-panel>
    </v-expansion-panels>
  </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';
import {
  mdiHeart,
} from '@mdi/js';

@Component
export default class Favourites extends Vue {
icons = { mdiHeart }
}
</script>
