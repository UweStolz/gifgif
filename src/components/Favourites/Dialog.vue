<template>
  <v-dialog
    v-model="syncedShowDialog"
    :full-image-data="syncedFullImageData"
    :image-id="syncedImageId"
    width="600"
    @click:outside="resetProps"
  >
    <v-card>
      <v-card-title>
        <v-spacer />
        <v-btn
          icon
          @click="resetProps"
        >
          <v-icon>{{ icons.mdiClose }}</v-icon>
        </v-btn>
      </v-card-title>
      <v-img
        :src="imageData"
        :lazy-src="imageData"
        max-height="700"
        height="100%"
        width="100%"
      />
      <v-card-actions>
        <v-btn
          :href="imageData"
          large
          icon
        >
          <v-icon
            id="downloadIcon"
          >
            {{ icons.mdiDownload }}
          </v-icon>
        </v-btn>
        <v-btn
          icon
          large
          @click="deleteGifData"
        >
          <v-icon
            id="trashIcon"
          >
            {{ icons.mdiTrashCan }}
          </v-icon>
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import {
  Vue, Component, PropSync, Watch, Emit,
} from 'vue-property-decorator';
import arrayBufferToImage from '@/util/imageHelper';
import {
  mdiClose, mdiDownload, mdiTrashCan,
} from '@mdi/js';

@Component
export default class Dialog extends Vue {
@Watch('syncedFullImageData')
  convertBufferToImage() {
    if (this.syncedFullImageData) {
      this.imageData = this.syncedFullImageData instanceof ArrayBuffer
        ? arrayBufferToImage(this.syncedFullImageData as ArrayBuffer)
        : this.syncedFullImageData as string;
    }
  }

@PropSync('showDialog', { type: Boolean, required: true }) syncedShowDialog!: boolean;

@PropSync('fullImageData', { required: true }) syncedFullImageData!: null|ArrayBuffer|string;

@PropSync('imageId', { required: true }) syncedImageId!: string;

@Emit('delete')
deleteGifData() {
  this.syncedShowDialog = false;
  return this.syncedImageId;
}

resetProps(): void {
  this.syncedShowDialog = false;
  this.syncedFullImageData = null;
}

icons = {
  mdiClose,
  mdiTrashCan,
  mdiDownload,
}

imageData: string = '';
}
</script>

<style scoped>
#trashIcon:hover{
  color: red;
}
#downloadIcon:hover{
  color:green;
}
</style>
