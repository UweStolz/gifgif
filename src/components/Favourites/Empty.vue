<template functional>
  <v-container
    fluid
    fill-height
  >
    <v-col
      class="text-center "
      cols="12"
    >
      <v-sheet
        class="display-2"
        tile
        width="100%"
        height="100%"
      >
        Nothing to show here ☹️
      </v-sheet>
    </v-col>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'Empty',
  functional: true,
});
</script>
