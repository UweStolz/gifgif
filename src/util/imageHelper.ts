export function base64ToGif(base64Image: string) {
  const image = new Image();
  image.src = `data:image/gif;base64,${base64Image}`;
  return image.src;
}

export function arrayBufferToBase64(buffer: ArrayBuffer): string {
  let binary = '';
  const bytes = new Uint8Array(buffer);
  const len = bytes.byteLength;
  for (let i = 0; i < len; i += 1) {
    binary += String.fromCharCode(bytes[i]);
  }
  const base64Image = window.btoa(binary);
  return base64Image;
}
