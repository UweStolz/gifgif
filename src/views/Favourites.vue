<template>
  <FavouritesView />
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import FavouritesView from '@/components/Favourites/index.vue';

 @Component({
   components: {
     FavouritesView,
   },
 })
export default class Favourites extends Vue {}
</script>
