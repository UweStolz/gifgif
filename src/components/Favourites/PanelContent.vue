<template>
  <v-expansion-panel-content id="pa-exp-con">
    <v-item-group>
      <v-row
        justify="start"
        align="start"
      >
        <Dialog
          :show-dialog.sync="dialog"
          :full-image-data.sync="fullImage"
          :image-id.sync="imageId"
          @delete="deleteGif"
        />
        <v-item
          v-for="(z, index) in previewImages.length"
          :key="z"
          class="mx-auto"
        >
          <v-lazy min-height="50">
            <v-img
              id="imageHover"
              :src="previewImages[index]"
              :lazy-src="previewImages[index]"

              max-width="230"
              max-height="200"
              @click="openDialog(index)"
            />
          </v-lazy>
        </v-item>
      </v-row>
    </v-item-group>
  </v-expansion-panel-content>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator';
import Dialog from './Dialog.vue';

@Component({
  components: {
    Dialog,
  },
})
export default class PanelContent extends Vue {
  @Prop({ required: true }) previewImages!: string[];

  @Prop({ required: true }) fullImages!: (Blob | string)[];

  @Prop({ required: true }) imageId!: string[];

  fullImage: null | Blob | string = null;

  dialog: boolean = false;

  selectedIndex: number = -1;

  openDialog(index: number) {
    this.dialog = true;
    this.fullImage = this.fullImages[index];
    this.selectedIndex = index;
  }

  async deleteGif(payload: string): Promise<void> {
    await this.$store.dispatch('removeGifData', payload[this.selectedIndex]);
    this.$store.commit('setGifCount', this.$store.state.gifCount - 1);
    this.previewImages.splice(this.selectedIndex, 1);
    this.fullImages.splice(this.selectedIndex, 1);
    this.imageId.splice(this.selectedIndex, 1);
    this.selectedIndex = -1;
  }
}
</script>

<style scoped>
#pa-exp-con {
  overflow: hidden;
}
#imageHover:hover {
  transform: scale(1.3);
  overflow-x: hidden;
  z-index: 999;
}
</style>
