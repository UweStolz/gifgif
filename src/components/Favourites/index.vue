<template>
  <div style="height: 100%;">
    <v-card-title
      v-if="!isDataEmpty && evaluatedKey !== 0"
      class="display-2 justify-center"
    >
      Favourites
    </v-card-title>
    <Empty v-if="isDataEmpty && evaluatedKey !== 0" />
    <v-expansion-panels
      v-if="!isDataEmpty && evaluatedKey !== 0"
      popout
      hover
    >
      <v-expansion-panel
        v-for="x in 5"
        v-show="data.previewUrl[x].previews.length === 0 ? false : true"
        :key="x"
      >
        <v-expansion-panel-header>
          <v-col cols="1">
            <v-avatar
              color="grey"
              size="32"
              v-text="data.previewUrl[x].previews.length"
            />
          </v-col>
          <v-row justify="center">
            <v-icon
              v-for="y in x++"
              :key="y"
              x-large
              color="red"
            >
              {{ icons.mdiHeart }}
            </v-icon>
          </v-row>
        </v-expansion-panel-header>
        <PanelContent
          :preview-images="data.previewUrl[x-1].previews"
          :full-images="data.previewUrl[x-1].images"
          :image-id="data.previewUrl[x-1].id"
        />
      </v-expansion-panel>
    </v-expansion-panels>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Watch } from 'vue-property-decorator';
import {
  mdiHeart,
} from '@mdi/js';
import blobToImage from '@/util/imageHelper';
import Empty from './Empty.vue';
import PanelContent from './PanelContent.vue';


interface Data {
  previewUrl: {
    [key: number]: { id: string[], images: (Blob | string)[], previews: string[] };
  }
}

@Component({
  components: {
    PanelContent,
    Empty,
  },
})
export default class Favourites extends Vue {
  @Watch('data', { deep: true })
  watchData() {
    let counter = 0;
    for (let index = 1; index <= 5; index += 1) {
      if (this.data.previewUrl[index].id.length === 0) { counter += 1; }
    }
    this.isDataEmpty = counter === 5;
  }

  icons = { mdiHeart }

  isDataEmpty: null | boolean = null;

  evaluatedKey: number = 0;

  data: Data = {
    previewUrl: {
      1: { id: [], images: [], previews: [] },
      2: { id: [], images: [], previews: [] },
      3: { id: [], images: [], previews: [] },
      4: { id: [], images: [], previews: [] },
      5: { id: [], images: [], previews: [] },
    },
  };

  async mounted(): Promise<void> {
    const gifData: Database.GifStore = await this.$store.dispatch('getAllData');
    this.isDataEmpty = gifData.keys.length === 0;
    this.evaluatedKey += 1;
    if (!this.isDataEmpty) {
      gifData.values.forEach((value: Database.GifData, index) => {
        if (value.rating > 0 && value.rating < 6) {
          let previewImageUrl = value.preview;
          if (value.preview instanceof Blob) {
            previewImageUrl = blobToImage(value.preview as Blob);
          }
          this.data.previewUrl[value.rating].previews.push(previewImageUrl as string);
          this.data.previewUrl[value.rating].images.push(gifData.values[index].image as any);
          this.data.previewUrl[value.rating].id.push(gifData.keys[index]);
        }
      });
    }
  }
}
</script>
