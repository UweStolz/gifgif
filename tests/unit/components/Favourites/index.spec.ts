import Favourites from '@/components/Favourites/index.vue';
import * as blobToImage from '@/util/imageHelper';
import shallow, { actions } from '../../../helper';

describe('Favourites', () => {
  it('renders properly', () => {
    const wrapper = shallow(Favourites);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Display no data if there is none', async () => {
    const wrapper = shallow(Favourites);
    await wrapper.vm.$nextTick();
    expect(wrapper).toHaveDispatched('getAllData');
    expect(wrapper.vm.$data.isDataAvailable).toBe(false);
  });

  it('Processes data, with previews as Strings', async () => {
    actions.getAllData.mockResolvedValueOnce(
      {
        values: [
          { rating: 1, image: 'image1.webp', preview: 'preview1.webp' },
          { rating: 2, image: 'image2.webp', preview: 'preview2.webp' },
          { rating: 10, image: 'image3.webp', preview: 'preview3.webp' },
        ],
        keys: ['key-1', 'key-2', 'key-3'],
      },
    );
    const wrapper = shallow(Favourites);
    await new Promise((resolve) => setTimeout(resolve));
    expect(wrapper).toHaveDispatched('getAllData');
    expect(wrapper.vm.$data.isDataAvailable).toBe(true);
    expect(wrapper.vm.$data.data.previewUrl[2].images[0]).toBe('image2.webp');
  });
  it('Processes data, with a preview as blob', async () => {
    jest.spyOn(blobToImage, 'default').mockImplementation(() => 'someImage.webp');
    actions.getAllData.mockResolvedValueOnce(
      {
        values: [
          { rating: 1, image: 'image1.webp', preview: new Blob() },
        ],
        keys: ['key-1'],
      },
    );
    const wrapper = shallow(Favourites);
    await new Promise((resolve) => setTimeout(resolve));
    expect(wrapper).toHaveDispatched('getAllData');
    expect(wrapper.vm.$data.isDataAvailable).toBe(true);
    expect(wrapper.vm.$data.data.previewUrl[1].previews[0]).toBe('someImage.webp');
  });
});
