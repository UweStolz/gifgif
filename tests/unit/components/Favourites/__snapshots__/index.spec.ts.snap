// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favourites renders properly 1`] = `
<div style="height: 100%;">
  <!---->
  <!---->
  <!---->
</div>
`;
